src/Microsoft.AspNetCore.Sockets.Common/Internal/Formatters/ServerSentEventsMessageParser.cs

```diff
@@ -0,0 +1,130 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.using System;
+
+using System;
+using System.Buffers;
+using System.IO.Pipelines;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.AspNetCore.Sockets.Internal.Formatters
+{
+    public  class ServerSentEventsMessageParser
+    {
+
+        public static ParsePhase ParseMessage(ReadableBuffer buffer, out ReadCursor consumed, out ReadCursor examined, out Message message)
+        {
+            const byte ByteCR = (byte)'\r';
+            const byte ByteLF = (byte)'\n';
+            const byte ByteSpace = (byte)' ';
+            bool firstLine = true;
+            byte[] data = null;
+            consumed = buffer.Start;
+            examined = buffer.End;
+            message = new Message();
+            MessageType messageType = MessageType.Text;
+            var reader = new ReadableBufferReader(buffer);
+
+            var start = consumed;
+            var end = examined;
+
+            while (!reader.End)
+            {
+                var span = reader.Span;
+                var backup = reader;
+                var ch1 = reader.Take();
+                var ch2 = reader.Take();
+                if (ch1 == ByteCR)
+                {
+                    if (ch2 == ByteLF)
+                    {
+                        consumed = buffer.End;
+                        message = new Message(data, messageType);
+                        return ParsePhase.Completed;
+                    }
+
+                    //When we are only missing the final \n
+                    return ParsePhase.Incomplete;
+                }
+
+                reader = backup;
+                if (ReadCursorOperations.Seek(start, end, out var lineEnd, ByteLF) == -1)
+                {
+                    // Not there
+                    return ParsePhase.Incomplete;
+                }
+
+                // Make sure LF is included in lineEnd
+                lineEnd = buffer.Move(lineEnd, 1);
+                var line = ConvertBufferToSpan(buffer.Slice(start, lineEnd));
+                reader.Skip(line.Length);
+
+                //Strip the \r\n from the span
+                line = line.Slice(0, line.Length - 2);
+
+                if (firstLine)
+                {
+                    messageType = GetMessageType(line);
+                    start = lineEnd;
+                    firstLine = false;
+                    continue;
+                }
+
+                var nothing = buffer.End;
+                //Slice away the 'data: '
+                var newData = line.Slice(line.IndexOf(ByteSpace) + 1).ToArray();
+                start = lineEnd;
+
+                if (data?.Length > 0)
+                {
+                    var tempData = new byte[data.Length + newData.Length];
```

Comment:
This is going to do a lot of copying for multi-line messages. Ideally, we'd want to simply keep returning `Incomplete` until we know we've got the whole frame, so that we can avoid copying until the end. Though that may be an optimization we add after we've gotten this code in. + @davidfowl for thoughts

For now though, let's avoid some of the extra copying. Use a `List<byte[]>` to hold the payload segments, and just add each line's `data` buffer to it. Then at the end, copy each one into a final buffer.